File: thallium-frontend/src/api/templates.ts
import { get, APIMissingTokenError } from "./client";
import store from "../store";

export interface Variant {
    variant_id: number;
    name: string;
    size: string;
    colour?: string;
    colour_code?: string;
    colour_code2?: string;
    price: string;
    last_synced: string;
}

export interface Template {
    template_id: number;
    title: string;
    product_id: number;
    mockup_file_url: string;
    last_synced: string;
    variants?: Variant[];
}

export const getTemplates = async (withVariants: boolean): Promise<Template[]> => {
    /* TODO: Handle lacking authorization */

    const token = store.getState().authorization.voucherToken;

    if (!token) {
        throw new APIMissingTokenError();
    };

    return await get(`/templates/?with_variants=${withVariants.toString()}`, {
        headers: {
            Authorization: `Bearer ${token}`,
        },
    }) as unknown as Template[];
};

File: thallium-backend/src/dto/templates.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel


class Template(BaseModel):
    """Base model for a template."""

    template_id: int
    title: str
    product_id: int
    mockup_file_url: str
    last_synced: datetime


class Variant(BaseModel):
    """Base model for a template."""

    variant_id: int
    name: str
    size: str
    colour: str | None
    colour_code: str | None
    colour_code2: str | None
    price: Decimal
    last_synced: datetime


class TemplateWithVariant(Template):
    """A voucher as stored in the database."""

    variants: list[Variant]
